### DISCUSSION TITLE
globalSetup setup functions run too fast in sequence

### DISCUSSION THREAD

**Discussion by paul-uz** [vitest-dev/vitest#5388] [Q&A]
I have 2 globalSetup files in the following order:

`globalSetup.ts`
```ts
import { ChildProcess } from 'child_process';
import DynamoDbLocal from 'dynamodb-local';

let teardownHappened = false;
let child: ChildProcess;

export async function setup() {
  child = await DynamoDbLocal.launch(8000, null, [], true, true);
}

export async function teardown() {
  if (teardownHappened) {
    throw new Error('teardown called twice');
  }
  DynamoDbLocal.stopChild(child);
}
```

`dynamodbSetup.ts`
```ts
import { CreateTableCommand, CreateTableCommandInput, DeleteTableCommand, DeleteTableCommandInput, DynamoDBClient, KeyType, ScalarAttributeType } from '@aws-sdk/client-dynamodb';

let teardownHappened = false;

const client = new DynamoDBClient({
  endpoint: 'http://127.0.0.1:8000',
  tls: false,
  region: 'eu-west-1',
  credentials: {
    accessKeyId: 'fakeMyKeyId',
    secretAccessKey: 'fakeSecretAccessKey',
  },
  maxAttempts: 50,
});

const tables = [
  {
    tableName: 'Table1',
    pkName: 'pk1',
    pkType: ScalarAttributeType.S,
    keyType: KeyType.HASH,
  },
  {
    tableName: 'Table2',
    pkName: 'pk2',
    pkType: ScalarAttributeType.S,
    keyType: KeyType.HASH,
  },
  {
    tableName: 'Table3',
    pkName: 'pk3',
    pkType: ScalarAttributeType.S,
    keyType: KeyType.HASH,
  },
];

export async function setup() {
  return await Promise.all(tables.map(async (table) => {
    const params: CreateTableCommandInput = {
      AttributeDefinitions: [
        {
          AttributeName: table.pkName,
          AttributeType: table.pkType,
        },
      ],
      KeySchema: [
        {
          AttributeName: table.pkName,
          KeyType: table.keyType,
        },
      ],
      ProvisionedThroughput: {
        ReadCapacityUnits: 1,
        WriteCapacityUnits: 1,
      },
      StreamSpecification: {
        StreamEnabled: false,
      },
      TableName: table.tableName,
    };
    console.debug(params);
    return client.send(new CreateTableCommand(params));
  }));
}

export async function teardown() {
  if (teardownHappened) {
    throw new Error('teardown called twice');
  }
  return await Promise.all(tables.map(async (table) => {
    const params: DeleteTableCommandInput = {
      TableName: table.tableName,
    };
    await client.send(new DeleteTableCommand(params));
  }));
}

```

The issue seems to be that `dynamodbSetup.ts` runs too quickly, before the local DynamoDB process has fully started. I tested this by running the tests and using `aws dynamodb list-tables --endpoint-url http://127.0.0.1:8000` to check the tables and it takes a few seconds to be able to connect.

How can I make sure the setup function in `globalSetup.ts` is complete?

**Comment by paul-uz** [comment]
So I actually increased the `maxAttempts` value to 100 and it seems to work, but only when running a single test. When I run all my tests, one test suite fails, as it cannot find the table, but it works when running that same test on its own :/

It's as if the tests start running before the globalSetup scripts finish?

**Comment by sheremet-va** [comment]
Global setup files are running one after another, if you need to wait a bit, you can add a `await timeout.promises.setTimeout(time)` in a setup function.

**Reply by sheremet-va** [reply]
Sorry, a little bit more clarification - setup files are imported in parallel, but the `setup` function is called in sequence, so you shouldn't connect to your database in a top level.

**Reply by paul-uz** [reply]
Hmm, I'll move some stuff about and see how it goes

**Reply by paul-uz** [reply]
Looks like the issue is that the command to load the database completes, but that does not indicate the database is ready to connect to. I will have to keep the large amount of retries. The other issue is that in the script that creates and deletes the tables, if i include the client in setup and teardown, its duplicating code that probably doesnt actually need doing, due to this timing issue.
